cpp/src/parquet/arrow/arrow_reader_writer_test.cc

```diff
@@ -5146,5 +5158,361 @@ TEST(TestArrowReadWrite, FuzzReader) {
+  ASSERT_OK_AND_ASSIGN(auto buffer, sink->Finish());
+
+  // Create reader to read page index.
+  auto read_properties = default_arrow_reader_properties();
+  auto reader = ParquetFileReader::Open(std::make_shared<BufferReader>(buffer));
+  auto metadata = reader->metadata();
+  ASSERT_EQ(1, metadata->num_row_groups());
+
+  // Make sure page index reader is not null.
+  auto page_index_reader = reader->GetPageIndexReader();
+  ASSERT_NE(page_index_reader, nullptr);
+  auto row_group_index_reader = page_index_reader->RowGroup(0);
+  ASSERT_NE(row_group_index_reader, nullptr);
+
+  // Setup expected data.
+  auto encode_int64 = [=](int64_t value) {
+    return std::string(reinterpret_cast<const char*>(&value), sizeof(int64_t));
+  };
+  const std::vector<std::string> c0_min_values = {encode_int64(1), encode_int64(3),
+                                                  encode_int64(6), ""};
+  const std::vector<std::string> c0_max_values = {encode_int64(2), encode_int64(4),
+                                                  encode_int64(6), ""};
+  const std::vector<std::string> c1_min_values = {"a", "c", "f", ""};
+  const std::vector<std::string> c1_max_values = {"b", "d", "f", ""};
+  const std::vector<int64_t> null_counts = {0, 0, 1, 2};
+  const std::vector<bool> null_pages = {false, false, false, true};
+  const size_t num_pages = 4;
+
+  // Verify page index by columns.
+  for (int col_idx = 0; col_idx < metadata->num_columns(); ++col_idx) {
+    // Verify column index.
+    auto column_index = row_group_index_reader->GetColumnIndex(col_idx);
+    ASSERT_NE(column_index, nullptr);
+    ASSERT_EQ(BoundaryOrder::Ascending, column_index->boundary_order());
+    ASSERT_EQ(num_pages, column_index->null_pages().size());
+    ASSERT_TRUE(column_index->has_null_counts());
+    for (size_t page_idx = 0; page_idx < num_pages; ++page_idx) {
+      ASSERT_EQ(null_pages[page_idx], column_index->null_pages()[page_idx]);
+      ASSERT_EQ(null_counts[page_idx], column_index->null_counts()[page_idx]);
+      if (col_idx == 0) {
+        ASSERT_EQ(c0_min_values[page_idx], column_index->encoded_min_values()[page_idx]);
+        ASSERT_EQ(c0_max_values[page_idx], column_index->encoded_max_values()[page_idx]);
+      } else {
+        ASSERT_EQ(c1_min_values[page_idx], column_index->encoded_min_values()[page_idx]);
+        ASSERT_EQ(c1_max_values[page_idx], column_index->encoded_max_values()[page_idx]);
+      }
+    }
+
+    // Verify offset index.
+    auto offset_index = row_group_index_reader->GetOffsetIndex(col_idx);
+    ASSERT_NE(offset_index, nullptr);
+    ASSERT_EQ(num_pages, offset_index->page_locations().size());
+    int64_t prev_offset = 0;
+    for (size_t page_idx = 0; page_idx < num_pages; ++page_idx) {
+      const auto& page_location = offset_index->page_locations()[page_idx];
+      ASSERT_EQ(static_cast<int64_t>(page_idx * 2), page_location.first_row_index);
+      ASSERT_GT(page_location.offset, prev_offset);
```

Comment:
my main question here is how do we know the actual offset here is correct (same with other tests)?

Reply:
Unfortunately we don't. We may need to change the `PageReader` to return a list of offsets to data pages, otherwise we have no idea. The only thing we are sure is that the page offsets are in the ascending order so here is why I use `ASSERT_GT` to check offsets.